File: src/connect4/ui/gui/swing/widgets/GUICoin.java
package connect4.ui.gui.swing.widgets;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;

import javax.swing.JPanel;

import connect4.controller.GameController;
import connect4.ui.gui.swing.events.GUICoinMouseListener;
import connect4.util.observer.IObserver;

/**
* @author:  Stefano Di Martino
* @created: May 27, 2012
*/

@SuppressWarnings("serial")
public class GUICoin extends JPanel implements IObserver {
	private final int RADIUS = 40;
	private final int column;
	private Color color;

	public GUICoin(final int column) {
		this.setCoinBackAndForegroundColor();
		this.color = Color.WHITE;
		this.column = column;
		this.addMouseListener(new GUICoinMouseListener(this));
	}

	private void setCoinBackAndForegroundColor() {
		this.setBackground(Color.BLUE);
		this.setForeground(this.color);
	}

	public void setColor(Color color) {
		this.color = color;
	}

	@Override
	public void paintComponent(Graphics g) {
		super.paintComponent(g); // paints background
		this.setCoinBackAndForegroundColor();

		Graphics2D g2 = (Graphics2D) g;

		// Activate Anti-Aliasing!
		g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
				RenderingHints.VALUE_ANTIALIAS_ON);
		g2.fillOval(18, 20, this.RADIUS, this.RADIUS);
	}

	public int getColumn() {
		return this.column;
	}

	@Override
	public void update() {
		GameController.getInstance().dropCoinWithSuccessFeedback(this.column);
	}
}

File: src/main/scala/connectfour/ui/gui/swing/widgets/GUICoin.scala
package connectfour.ui.gui.swing.widgets

import java.awt.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.RenderingHints
import javax.swing.JPanel
import connectfour.ui.gui.swing.events.GUICoinMouseListener
import connectfour.ui.gui.swing.events.MouseColumnObserver
import connectfour.util.observer.IObserver
import connectfour.util.observer.IObserverWithArguments
import scala.beans.BeanProperty
import connectfour.controller.Connect4GameController

class GUICoin(controller: Connect4GameController, column: Int, observer: IObserverWithArguments) extends JPanel with IObserver {
  val RADIUS = 40;
  val X_POSITION = 18;
  val Y_POSITION = 20;
  
  @BeanProperty
  var color = Color.WHITE

  {
    this.setCoinBackAndForegroundColor
    this.addMouseListener(new GUICoinMouseListener(this));
    this.addMouseListener(new MouseColumnObserver(observer, column));
  }

  def setCoinBackAndForegroundColor {
    this.setBackground(Color.BLUE);
    this.setForeground(this.color);
  }
  

    
    override def paintComponent(g: Graphics) {
        super.paintComponent(g); // paints background
        setCoinBackAndForegroundColor
        
        val g2 = g.asInstanceOf[Graphics2D]
        
        // Activate Anti-Aliasing!
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        val radius = RADIUS;
        g2.fillOval(X_POSITION, Y_POSITION, radius, radius);
    }
    
    
    override def update() {
    	controller.dropCoin(column);
    }
}